app/models/main.py

```diff
@@ -1319,6 +1320,16 @@ class Brief(db.Model):
         viewonly=True
     )
 
+    process_outcome = db.relationship(
+        "ProcessOutcome",
+        primaryjoin=lambda: (sql_and(
+            foreign(Brief.id) == remote(ProcessOutcome.brief_id),
```

Comment:
Seven lines above this we use the string notation to define a primary join. Here we pass it to a lambda.
We should pick one way.

Reply:
The lambda is only required when we're referencing models "lower down" in the file (i.e. haven't been declared in this scope yet), but using it where it's not required would just be line noise.

Reply:
That's the same reason we're using a string 7 lines above. 
```
        primaryjoin="_and(Brief.id==BriefResponse.brief_id, BriefResponse.awarded_at.isnot(None))",
```

Reply:
The string method is slightly nastier & more vague (what are the scoping rules for those strings? what operators & functions are imported?). They also thwart the ability for "smart" IDEs that some people use to track variables or give comtextual information. If there are string-based clauses above written by *me*, it's because I didn't at that point know the lambda trick was supported.

Reply:
When we pick a new way to do things it's usually worth updating existing instances at the same time. Otherwise we're increasing tech debt and inconsistency.

Alternatively, if that sounds too long, we can do the old way and put a tech debt ticket up to update to the new way. That way at least we're not increasing inconsistency.

Putting up these smaller tech debt tickets is really helpful for new starters and junior devs who need to get to know the codebase and need to be able to get stuck in quickly. Having a few small tickets like going through and updating ORM definitions provides exposure to the codebase and builds confidence.

The inconsistency might not throw you off because you know the codebase, in the same way that going through the model layer and updating all the string relationship definitions to lambdas sounds like a chore, but to a new starter or junior they are respectively a real problem and an opportunity.

Reply:
And every extra line of code I unnecessarily touch is another thing that someone will potentially pull me up for, maybe the indentation, or potentially misunderstand what I'm doing and make a complaint about something that was in the original code that is being rewritten, or maybe stop everything and want to organize a meeting about "which way are we going to do this?". On the other hand *cautiously* doing things in "new ways" allows us to experiment and learn which ways are better or worse without selling the farm.

Reply:
I don't agree with this at all when you're coding in a team. You're saying that not committing to a single good way of doing something is a good way to introduce changes?

Reply:
It's not that I think that lambda > strings or vice versa it's that we should be pushing towards a more consistent codebase rather than less. 

If you're in a file and all the methods are in camelCase then you'd use camelCase, if everything is written in functions rather than classes then use functions, if strings are interpolated with `'%s' % var` instead of `'{}'.format(var)` use the former. 

This comment is not at all about which way is better. It's about picking a way and sticking to it because consistency is valuable in itself.

Inconsistent code is harder to read and harder to trust. If you're reading a file and 10% of the methods are in snake case and 90% in camel it feels... off? It throws you off, it doesn't quite look professional. Do these functions work together? Did the writers know where they were going with this? Did they know what eachother were doing? Why did they do it differently? Is there some sort of pitfall I'm missing here?

Put differently I wouldn't care if a python project used camel case for all of it's variables, as long as it did it consistently. It's not hurting anything. But if you want to start using snake case then convert the whole project. Otherwise you have a legacy code and new code split. New devs have to ask 'What are we up to here? Snake or camel? Why the split?'
What about if you treat every decision like that? Functions over classes? A Snake Function/ Camel Function/ Snake Class/ Camel Class split? Then you want to do string interpolation a new way so you have a Old String Snake function Old string Camel function New String Camel Snake exponential increase 🤷‍♂️ 

My only point is that doing this consistently would make this PR better. 🤷‍♂️

Reply:
C'mon.. the lettercase thing is a something of a straw man argument: it's not the scenario we're considering. I absolutely agree that we have to be consistent in a great many things including in how we name/case our variables/functions/API keys.

But here we're just saying that a library offers multiple ways to do a thing, and that's true all over with Python -- there's frequently multiple ways to do something, and sometimes one way may be more clear/neat/expressive and sometimes the other, and the tradeoffs may be different in different parts of the codebase, and frequently they are subjective anyway.

If we really want random unrelated counterexamples, then *for example* I wouldn't agree with a rule that said "all lambda's are banned because you could write them consistent with other functions and give them a name"! - what should be consistent is a matter of judgement and I think you and Rob might just need to agree to differ here ;-)